src/snowflake/connector/cache.py

```diff
@@ -593,14 +604,17 @@ def _should_load(self) -> bool:
 
     def __del__(self) -> None:
         try:
-            self._save()
+            with self._lock:
+                if self._cache_entry_updated:
+                    self._save()
         except Exception:
             # At tear-down time builtins module might be already gone, ignore every error
             pass
 
     def clear_expired_entries(self) -> None:
         super().clear_expired_entries()
-        self._save_if_should()
+        with self._lock:
+            self._save_if_should()
```

Comment:
Since you made `_should_save` rely on a self attribute please add to the doc-strings of the functions that call it that it's expected that you're holding `self._lock` when calling them. This would affect `_save_if_should` at least.